File: src/main/java/com/wesker926/leetcode/algorithms/p0430/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/12/18
// @description: TODO
package s1

type Node struct {
	Val   int
	Prev  *Node
	Next  *Node
	Child *Node
}

func flatten(root *Node) *Node {
	dfs(root)
	return root
}

func dfs(root *Node) *Node {
	var last, t *Node
	for ; root != nil; last, root = root, root.Next {
		if root.Child == nil {
			continue
		}

		t = dfs(root.Child)
		t.Next = root.Next
		if t.Next != nil {
			t.Next.Prev = t
		}
		root.Next = root.Child
		root.Child.Prev = root
		root.Child = nil
		root = t
	}
	return last
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0430/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0430.s1;

/**
 * @author wesker.gh
 * @date 2021/12/18
 * @description dfs
 */
public class Solution {
    public Node flatten(Node head) {
        dfs(head);
        return head;
    }

    private Node dfs(Node head) {
        Node last = head, t;
        for (; head != null; last = head, head = head.next) {
            if (head.child == null) {
                continue;
            }

            t = dfs(head.child);
            t.next = head.next;
            if (t.next != null) {
                t.next.prev = t;
            }
            head.next = head.child;
            head.child.prev = head;
            head.child = null;
            head = t;
        }
        return last;
    }

    private static class Node {
        public int val;

        public Node prev;

        public Node next;

        public Node child;
    }
}
